config/settings.yml

```diff
@@ -1077,10 +1077,12 @@
   :worker_base:
     :defaults:
       :count: 1
+      :cpu_minimum_percent: 25
       :cpu_threshold_percent: 50
       :gc_interval: 15.minutes
       :heartbeat_freq: 10.seconds
       :heartbeat_timeout: 2.minutes
+      :memory_minimum: 250.megabytes
```

Comment:
As said elsewhere, I'm not sure we should specify minimum defaults.

Reply:
> As said elsewhere, I'm not sure we should specify minimum defaults.
> -- @Fryguy 

yes. and that is why I had the first set of code. I avoided setting a default request for all workers.

But when @jrafanie requested I set defaults in the settings, I realized that while we didn't want to set the request by default, we were setting the limit by default.  So indirectly, we were always setting a request (minimum memory needed).

> Here's the configuration file for a Pod that has one Container. The Container specifies a memory limit, but not a request:
> [...]
>The output shows that the Container's memory request is set to match its memory limit. Notice that the Container was not assigned the default memory request value of 256Mi.
> [ref] 

If we want it to work with lower constraints then we have a few options:

- turn off resource limits as a whole
- not set limits to all workers by default
- change our code to set request where ever a limit is set

I felt like we going with the last option.
And since we are always setting a limit, we are always setting a request.

I feel like I may have drawn a faulty conclusion. Specifically around the fact that setting a limit also sets a request. I saw no such reference in the [OpenShift docs]

[ref]: https://kubernetes.io/docs/tasks/administer-cluster/manage-resources/memory-default-namespace/#what-if-you-specify-a-container-s-limit-but-not-its-request
[OpenShift docs]: https://docs.openshift.com/container-platform/3.11/dev_guide/compute_resources.html

Reply:
4th option... honor nil/0 as a valid value that means no limit.  Having a place already in the settings that we can easily change is helpful to avoid copy/paste/typos.

Reply:
> So indirectly, we were always setting a request (minimum memory needed).

Oh wow...that is completely unexpected to me...  We intentionally chose to not set it so it _wouldn't_ do that

Reply:
> enforce_resource_constraints

I think enforce_resource_constraints was a hack option until we got the individual request/limit separation...I think ultimately we should drop that option and always enforce it.

> So indirectly, we were always setting a request (minimum memory needed).

I think I misinterpreted the way it worked and that the default of `enforce_resource_constraints=false` was effectively _not_ setting the limit and therefore not setting the _request_.  However when `enforce_resource_constraints` was set to true, then yes, it seems to set both limit and request to match.

Reply:
After all that, I am :+1: on `change our code to set request where ever a limit is set`

Reply:
>I think ultimately we should drop [encorce_resource_constraints] and always enforce it.

yea, lets drop `enforce_resource_constraint` when this is figured out

this is fascinating. 

the [openshift] docs do a good job defining the various parameters. They do not mention the default request getting overridden when a limit is specified, but they do talk about min, max, request, and default (think default_max would be a better name here), and defaultRequest.

[openshift]: https://docs.openshift.com/container-platform/3.11/dev_guide/compute_resources.html#dev-viewing-limit-ranges

I think we may want to stop stating a limit in the default worker.